sdk/core/Azure.Core.Experimental/src/JsonData.cs

```diff
@@ -286,116 +196,61 @@ public override bool Equals(object? obj)
         }
 
         /// <inheritdoc />
-        public bool Equals(JsonData other)
+        public bool Equals(JsonData? other)
         {
-            if (_kind != other._kind)
+            if (other is null)
             {
                 return false;
             }
 
-            switch (_kind)
+            if (Kind != other.Kind)
             {
-                case JsonValueKind.Null:
-                case JsonValueKind.Undefined:
-                    return true;
-                case JsonValueKind.Object:
-                    return _objectRepresentation!.Equals(other._objectRepresentation);
-                case JsonValueKind.Array:
-                    return _arrayRepresentation!.Equals(other._arrayRepresentation);
-                default:
-                    return _value!.Equals(other._value);
+                return false;
             }
+
+            return _element.Equals(other._element);
         }
 
         /// <inheritdoc />
-        public override int GetHashCode()
-        {
-            if (_kind == JsonValueKind.String)
-            {
-                return ((string?)_value)!.GetHashCode();
-            }
+        public override int GetHashCode() => _element.GetHashCode();
 
-            return base.GetHashCode();
-        }
+        private string? GetString() => _element.GetString();
 
-        private string? GetString() => (string?)EnsureValue();
+        private int GetInt32() => _element.GetInt32();
 
-        private int GetInt32()
-        {
-            var value = EnsureNumberValue().AsLong();
-            if (value > int.MaxValue || value < int.MinValue)
-            {
-                throw new OverflowException();
-            }
-            return (int)value;
```

Comment:
This means we won't get `InvalidCastExceptions` anymore, but will instead standardize on exceptions thrown by JsonElement, which is nice for consistency with BCL.

Reply:
But throwing STJ exceptions is leaking implementation details.

Reply:
Tracking with https://github.com/Azure/azure-sdk-for-net/issues/33856